Lib/glyphsLib/classes.py

```diff
@@ -1747,6 +1760,91 @@ def _set_metric(self, metricname, value):
             self.metrics.append(GSMetricValue(position=0))
         self.metrics[metricIndex] = GSMetricValue(position=value)
 
+    @property
+    def alignmentZones(self):
+        # If there are values that are parsed from file format 2 or which are
+        # explicitly set return those
+        if self._alignmentZones is not None:
+            return self._alignmentZones
+
+        if len(self.metrics) == 0:
+            return []
+
+        validZoneMetrics = (
+            "ascender",
+            "cap height",
+            "x-height",
+            "baseline",
+            "descender",
```

Comment:
where's this list of validZoneMetrics taken from? it seems constant, it should probably be defined at global scope, or in the constants module.

Reply:
and why skip some of the font metric types as not "valid" when fetching a given master metric values? where is this logic encoded?

Reply:
Basically I went by what's in the UI (mind you I missed `bodyHeight`):

<img width="501" alt="Screenshot 2022-12-16 at 21 09 41" src="https://user-images.githubusercontent.com/1893577/208174153-30b8b9ea-26ac-467e-9d4c-d9571ceb57a2.png">

Outputting the `GSMetric.type` of these objects of a file gives:

```
ascender
cap height
x-height
baseline
descender
italic angle
bodyHeight

```

The last empty line for type is what the UI has as "Custom" metric. This may or may not be an overshoot — presumably things with overshoot != 0 we want to treat as an alignmentZone as well?

I think when I did a "has no overshoot" exclusion instead of explicitly checking for the type _a lot_ of tests outright failed because the font metricValues does not match a given master's metricValues length (e.g. in all sorts of constructed test cases or programmatic access of those objects that don't adhere to the file format 3 referencing the font metricValues from the master by index). As far as I recall the `"baseline": 0` default above also addressed this.

Note sure what to do here, but I'll ponder it.

Reply:
Updated in code below in forced push. Disregarding the "italic angle" type of metrics early was key to avoiding all those issues in tests that construct just masters. Gets rid of the explicit magic list and does not require the `baseline` default either. Thanks for poking holes @anthrotype :)

Reply:
> ... because the font metricValues does not match a given master's metricValues length (e.g. in all sorts of constructed test cases or programmatic access of those objects that don't adhere to the file format 3 referencing the font metricValues from the master by index

if it's invalid that len(master.metricValues) != len(font.metrics) then it is those tests that would need to be fixed...

Reply:
The tests were correct, only when _not filtering_ things that are now returned by a master'a metrics which are not alignmentZones in format 3 those test broke because without knowledge of the master my previous iteration of the code could not filter those properly. Different approach in the updated code, essentially discarding the `italic angle` from being an alignmentZone outright, and further removing metrics which do not have an overshoot, e.g. do not form a bluezone because they would be a "line" only.

Reply:
can a valid "italic angle" metric ever have overshoot != 0? I suppose not. And if not, then simply filtering out metrics with overshoot=0 (without special-casing italic angle itself) should accomplish the same thing.

Reply:
In
```
 for index, fontMetric in enumerate(self.font.metrics):
            # Ignore the "italic angle" "metric", it is not an alignmentZone
            if fontMetric.type == "italic angle":
                continue
            metric = self.metrics[index]
            # Ignore metric without overshoot, it is not an alignmentZone
            if metric.overshoot == 0:
                continue
            zone = GSAlignmentZone(pos=metric.position, size=metric.overshoot)
            zones.append(zone)
```

`self.font.metrics` is a [GSMetric](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/Lib/glyphsLib/classes.py#L1393) value which has no overshoot. One has to use the index of this  _font_ metric values to get the _master_ [GSMetricValue](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/Lib/glyphsLib/classes.py#L1408) value, and then discard things which have no overshoot. Filtering the _font_ metric values by type `italic angle` is necessary because just accessing `master.metrics[]` by index will not have a value for a _font_ metric that is an italic angle.

The format 3 handling of metric values referencing back up to the font metrics list is convoluted imo.